### PR TITLE
Feature/tutor create form

### LINKED ISSUE #1432: Create user (Tutor) form
### Link task to story

Story: https://github.com/Open-Study-College/osc/issues/1419

### Description

Name, email, list of courses, IV user (yes or no)

### LINKED ISSUE #1433: Hook form up to action/create user/
### Link task to story

Story: https://github.com/Open-Study-College/osc/issues/1419

### Description

Registration should send magic email link which will redirect tutor back to page, so they can create their password.

### PR DESCRIPTION
**Pull request by stevenbister**
### 🧹 Please check off the following before submitting:

- [x] Close issues
- [x] Request reviewers
- [x] Assign yourself
- [ ] Add labels (e.g. `build:osc-ecommerce`)
- [x] Add project
- [x] Set project status
- [x] Add milestone

---

## ☑ List all github issues that this PR will close:

- Closes #1432 
- Closes #1433 

## 📝 Description

Adds the tutor create form to Ohana. This will sit on the `/tutors/create` route which is protected by the user's permissions, if a user lands on this page without permissions it will throw a 403.
The action will check that the user is authenticated and refresh the token if needed before using that token as part of the create tutor function.

I've added a `SearchableCheckboxListModal` component which will allow users to search from a list of checkboxes and select and save the value they want, it's fully frontend at the moment and will need to be integrated with the search endpoint once it's ready. Have added a comment in the file and opened a ticket: #1609.

I'm hoping it will look better once #1526 is merged as at the moment it is too wide, and the background colour is a bit strong, if needed I'll make adjustments to it then.

## ⛳️ Current behaviour (updates)

N/a

## 🚀 New behaviour

- Adds a `tutors/create` route
- Adds `CreateTutorForm` component
- Adds `SearchableCheckboxListModal`
- Some tweaks to osc-ui components
- Adds utility functions to check user permissions

## 💣 Is this a breaking change (Yes/No): No

> Please describe any breaking changes.

## 🧾 Final Checklist

- [ ] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [x] I've considered possible cross-device impacts (iOS, Android)
- [x] I've considered possible cross-browser impacts (Chrome, Safari, Edge, Firefox)
- [x] I've added tests to cover my changes (Unit, E2E, Integration, Contract, Functional, Acceptance etc)
- [ ] I've added or updated any documentation.

## 📝 Additional Information

This actually won't work if you try submitting as we are waiting for a couple of bugfixes which are in #1596. If you want to test before this is merged you can add the contents of https://github.com/Open-Study-College/osc/pull/1596/commits/2632590db4e5c1d6677737b5526b1786bf76302f to get it working

Should also mention there might be a few ts errors / comments, which will get fixed when #1596 goes in.

### Summary by CodeRabbit

- New Feature: Added a new `CreateTutorForm` component for creating tutors and allocating courses to them.
- New Feature: Introduced a `SearchableCheckboxListModal` component that provides a modal with a searchable list of checkboxes.
- New Feature: Implemented new GraphQL queries and mutations for fetching course data, creating tutors, and managing user permissions.
- Refactor: Updated the `ButtonGroup` component in the UI library to support different justification options.
- Test: Added comprehensive test suites for the new components and utility functions.
- Chore: Updated various type definitions and validation schemas to accommodate the new features.

### CODE REVIEW

File: packages/osc-academic-hub/app/components/Registration/CreateTutorForm.tsx [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+                        label="Email"
+                        type="email"
+                        variants={['secondary']}
+                        errors={fieldErrors?.email}
+                        setErrors={setFieldErrors}
+                        schema={createTutor.pick({ email: true })}
+                    />
+                </div>
+
+                <div className="o-grid__col o-grid__col--12">
```
[the remark this answers is not in the document]
Comment (stevenbister) [reply packages/osc-academic-hub/app/components/Registration/CreateTutorForm.tsx:84]: Our `Checkbox` component has a label built into it so we don't need to specify one outside of it

File: packages/osc-academic-hub/app/utils/courses.spec.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+                    id: 2,
+                    title: 'Test Course 2',
+                },
+                {
+                    id: 3,
+                    title: 'Test Course 3',
+                },
+            ]),
+        };
+    });
```
[the remark this answers is not in the document]
Comment (stevenbister) [reply packages/osc-academic-hub/app/utils/courses.spec.ts:22]: This is incorrect, we have vitest installed which uses `vi` not `jest`

File: packages/osc-academic-hub/app/utils/courses.spec.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+});
+
+afterEach(() => {
+    vi.clearAllMocks();
```
[the remark this answers is not in the document]
Comment (stevenbister) [reply packages/osc-academic-hub/app/utils/courses.spec.ts:26]: This is incorrect, we have vitest installed which uses `vi` not `jest`

File: packages/osc-academic-hub/app/components/Registration/CreateTutorForm.tsx [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+            {error && selectedCourses.length === 0 ? (
+                <p className="u-color-error">{error}</p>
+            ) : null}
+
+            <input
+                type="hidden"
+                name="courses"
+                value={selectedCourses.length ? JSON.stringify(selectedCourses) : ''}
+            />
```
[the remark this answers is not in the document]
Comment (stevenbister) [reply packages/osc-academic-hub/app/components/Registration/CreateTutorForm.tsx:168]: While I do agree it's probably not best to send so much data, we need the shape of the object when handling it in our server action

File: packages/osc-academic-hub/app/routes/tutors.create.tsx [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+            });
+        }
+
+        // Thrown error will be caught by route error boundary
+        throw json<LoaderData>(
+            { error: 'User does not have permission: `createOtherAccount`' },
+            { status: 403 }
+        );
+    }
+};
```
[the remark this answers is not in the document]
Comment (stevenbister) [reply packages/osc-academic-hub/app/routes/tutors.create.tsx:119]: This is fine, courses aren't protected by permissions and can be viewed by anyone

File: packages/osc-academic-hub/app/components/SearchableCheckboxListModal/SearchableCheckboxListModal.tsx [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+
+export const SearchableCheckboxListModal = (props: SearchableCheckboxListModalProps) => {
+    const { triggerLabel, modalTitle, searchLabel, items, defaultSelectedItems, setExternalState } =
+        props;
+
+    const [isOpen, setIsOpen] = useState<boolean>(false);
+    const [searchValue, setSearchValue] = useState<string>('');
+    const [selectedItems, setSelectedItems] = useState<typeof items>([]);
+    const [filteredItems, setFilteredItems] = useState<typeof items>(items);
+
```
[the remark this answers is not in the document]
Comment (stevenbister) [reply packages/osc-academic-hub/app/components/SearchableCheckboxListModal/SearchableCheckboxListModal.tsx:60]: We are initialising with an empty array yes, but we're then using a `useEffect` to set the state when the `defaultSelectedItems` changes. This does a better job of keeping the external and internal states in sync

File: packages/osc-academic-hub/app/utils/user.spec.ts [hunk 7]
```diff
@@ -28,3 +29,87 @@ test('returns the user data', async () => {
...
+                        write: ['createOtherAccount'],
+                    },
+                },
+            } as User;
+
+            expect(hasPermission(user, 'canReadOtherAccount')).toBe(false);
+            expect(hasPermission(user, 'canEditCourseContent')).toBe(false);
+        });
+    });
+});
```
[the remark this answers is not in the document]
Comment (stevenbister) [reply packages/osc-academic-hub/app/utils/user.spec.ts:115]: The user will never be null or undefined as typescript will catch that before the data goes into the function. However Adding a test for the empty arrays is a good shout

**Review by mattwendzina — COMMENTED** [review]
Nice mate, all seems to work well 👍 
LGTM

**Review by Tom-Pearce — COMMENTED** [review]
Not looking at code, just testing functionality.

We should probably redirect the user _away_ from the `tutor/create` once they submit the form. Alternatively offer them a second submit button labelled something like `create and add another` or something. Alternatively remain on the page hide the form and show the confirmation message alongside a "create another" button that resets the form.

Also the terminology on this page is a bit mixed. The route is `tutors/create` but the text all says "add", I would probably stick to `create` as `add` to me indicates linking 2 existing records together, e.g. you might add a course (it already exists) to a tutor, but `create` means to create a new record, e..g create the tutor user record.

**Review by Tom-Pearce — APPROVED** [review]
Functionality LGTM 🚀
